src/provider/utils/async.ts

```diff
@@ -81,6 +91,39 @@ export function allowReject<T>(promise: Promise<T>): Promise<T> {
     return promise;
 }
 
+/**
+ * Takes multiple promises representing API calls, and reduces them to a single result. In the case that more than one promise resolves to
+ * a result, the first to result will be used. Intented to be used when calling a client from the provider, to protect against a
+ * misbehaving client.
+ *
+ * @param promises An array of promises
+ */
+export async function collateApiCallResults<T = void>(promises: Promise<T>[]): Promise<[CollateApiCallResultsResult, T | undefined]> {
```

Comment:
If this is going to be in the async utils, can it be a bit more generic. I.e. not explicitly referencing API calls or externally defined timeouts. I expect we'll be making this a standalone module soon and would prefer to keep it as independent as possible for when that day comes.

Reply:
In that case I'd rather go the other direction - keep this specific to API calls, but move it somewhere other than `async`. (And if it makes sense, pull out a generic part of the implementation into `async`). Happy for me to take that approach?

Reply:
Yea, that's fine. My metric would be anything which would be useful to any random block of promises should stay here, and anything that is only really useful in the context of the services belongs elsewhere.

Reply:
✅

Moved into `provider/utils/helpers.ts`